Allow collections to define their own action groups

s-hertel opened the pull request:
##### SUMMARY
Fixes #59864

Groupings should be in a collection's meta/action_groups.yml file in the format:
```
group_name:
  - module_name
```

##### ISSUE TYPE
- Feature Pull Request

s-hertel commented on lib/ansible/executor/module_common.py:
```diff
@@ -1288,7 +1289,10 @@ def modify_module(module_name, module_path, module_args, templar, task_vars=None
 
 
-def get_action_args_with_defaults(action, args, defaults, templar):
+def get_action_args_with_defaults(action, args, defaults, templar, collection_list=None):
```
may be better to just pass task at this point

bcoca replied:
yep, but tempted to just wait and revamp the whole system, this shoudl just give us a transition to collections

bcoca commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -19,14 +19,23 @@
+    # use C version if possible for speedup
+    from yaml import CSafeLoader as SafeLoader
+except ImportError:
+    from yaml import SafeLoader
```
tempted to use dataloader .. but that might enable json and vaults

s-hertel commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -330,3 +333,22 @@ def parse(self, skip_action_validation=False):
+    if len(name.split('.')) == 6:
+        collections_dir, collection_namespace, collection_name, section, subsection, shortname = name.split('.')
+        dir_list = action_path.split(os.path.sep)[0:-3]
+        action_groups_path = os.path.join(os.path.sep.join(dir_list), 'meta', 'action_groups.yml')
```
I should probably handle .yml | .yaml. Should I add json ?

nitzmahone commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -330,3 +333,22 @@ def parse(self, skip_action_validation=False):
+
+
+def get_action_groups_config(action, collection_list):
+    name, action_path = module_loader.find_plugin_with_name(action, collection_list=collection_list, check_aliases=True)
```
It looks like you're assuming the group lives in the same collection as the task action- it might not. The `module_defaults` declaration can be one of:
* `group/(FQCN of group name)` - eg, `group/community.aws.aws`
* `group/(unqualified group name)` - eg `group/aws`
* `(action/module FQCN)` - eg `community.aws.ec2`
* `(unqualified action/module name)` - eg `ec2`

The first case is straightforward- ask the named collection's metadata for the action group def, error if not found, party on. The second case should walk the collection list and ask each one for the group name, stopping at the first found definition (and later, once collection redirection is merged, consult the built-in unqualified redirect list before giving up). The third and fourth cases are basically just dynamic un-named groups of a single action/module.

This also points out to me that the lookup of the action groups and the storage of the defaults for the task probably belong in the playbook/DS hierarchy validation. That way, the metadata consultation is done in the controller (and the group metadata can be cached easily there), instead of repeatedly parsing and consulting the metadata in the workers.

s-hertel replied:
The second one feels kind of odd. I should expect that collections can use each other's action_groups? That seems like it would be a surprise to me as a user and would rather have things safely namespaced. Why not just inspect the collection of the task action and if it's not there, there is no group? The 1st's syntax to me seems pretty unfortunate but perhaps better than what I've done practically. - Edit: thinking about this some more, yes, I have to add the first syntax. Otherwise group names in different collections would get the same module_defaults arguments and would have the namespacing problem again. I'd question now whether group/(unqualified group name) should even be allowed, except that it would break current playbooks.

I did specifically avoided touching the action/module stuff since this PR was just to fix groupings, but I have looked into it and played with some possible solutions. FQCN works currently, except for ansible.builtin and ansible.legacy, which are just broken.

I asked bcoca about caching the results to improve performance and decided to hold off until I had something working decently. And yes, let's talk about the playbook/DS - the placement of this stuff feels a little awkward.

nitzmahone reviewed (commented):
Let's chat on this one- I think in order to do this in a performant manner, we need to move a little more stuff into the playbook/DS hierarchy validation rather than handling it post-fork in the worker. With a single shared config as it was originally, it made more sense where it is, but probably not anymore.

There are also a couple of things coming in the collection meta/redirection PR that you could leverage to get rid of a lot of code (eg we could just autoload the action_groups on initial collection load during playbook parse, and index them for snappy action lookups).

s-hertel committed:
Everything is FQCN for simplicity: actions in the playbook, actions listed in action_groups, groups in module_defaults

s-hertel committed:
Allow abbreviated names for task actions to work with action groups

s-hertel committed:
Allow abbreviated module names in action_defaults.yml which will be used for the current collection

s-hertel committed:
Fix a little bug and comment code where there are still changes to be made

Note: group/namespace.collection.module is supported currently, group/module is waiting on redirection changes.

s-hertel committed:
WIP integration tests - use real collections instead

geerlingguy commented:
Will there be any documentation on the format / purpose of the `action_groups` file? Is this something that's only really required for modules that are being migrated out from subdirs in ansible/ansible to collections, or is this something required (and/or optional?) for all collections?

s-hertel commented:
@geerlingguy I will add documentation. The groups can be helpful for actions/modules that have a common set of options. For example, all of the AWS modules that take the same parameters for authentication are in the group `aws`, so instead of providing those for each task the default values for the modules in those could be set just once via module_defaults:
```
module_defaults:
  group/ansible.amazon.aws:
    aws_access_key: "{{ access_key }}"
    aws_secret_key: "{{ secret_key }}"
    aws_security_token: "{{ token }}"
    region: us-east-1
```
Here's the documentation for non-collection module_defaults groups: https://docs.ansible.com/ansible/latest/user_guide/playbooks_module_defaults.html#module-defaults-groups

tremble commented:
> ```
> module_defaults:
>   group/ansible.amazon.aws:
> ```

I don't know how many examples there will be of this, but with the Community/Core split on the AWS side it would be nice **not** to _require_ people to use two separate group namespaces.  At the least we should allow the 'community' modules to use the 'core' group.

s-hertel commented:
@tremble I agree. That example was really just to display the new syntax. `group/aws` will keep working for the things that have worked before. There was a comment in IRC that new things should only work with FQCN groups (which was not on my radar when I was working on this), but unless it's actively vetoed I plan on keeping `group/aws` working for new modules as well.

s-hertel committed:
Update location of action_groups in meta/runtime.yml

Add action_groups_redirection to support unqualified group names

s-hertel committed:
Add some user and developer documentation

s-hertel committed:
Fix sanity tests

Update unit tests

Remove cruft from integration test

Add ansible_internal_redirect_list as a valid keyword for TaskInclude

s-hertel committed:
Remove hardcoded module_defaults.yml

remove the tests that verified the hard coded module_defaults.yml file

felixfontein commented on lib/ansible/executor/module_common.py:
```diff
@@ -1331,14 +1335,36 @@ def get_action_args_with_defaults(action, args, defaults, templar):
+
+                tmp_args.update((module_defaults.get('group/{0}'.format(fqcr)) or {}).copy())
+
+                redirected_groups = redirect_action_groups.get(fqcr, redirect_action_groups.get(group, []))
+
+                for redirected_group in redirected_groups:
+                    tmp_args.update((module_defaults.get('group/{0}'.format(redirected_group)) or {}).copy())
```
When I look at your above example:
```.yaml
   action_groups_redirection:
     another_group:
       redirect: exact_match
```
then `redirected_groups` would be a `dict`, and `redirected_group` will be `redirect`. If I use
```.yaml
   action_groups_redirection:
     another_group:
       - exact_match
```
instead, it works.

s-hertel replied:
Fixed

felixfontein commented on docs/docsite/rst/dev_guide/developing_collections.rst:
```diff
@@ -195,6 +197,92 @@ command completion, or environment variables are presented throughout the
+
+     action_groups_redirection:
+       group_name:
+         redirect: common_group
```
Here is the documentation I'm referring to.

s-hertel replied:
The documentation was correct, code was wrong

felixfontein reviewed (approved):
Did some testing, seems to work fine!

gundalow commented:
validation of `runtime.yml` done in https://github.com/ansible/ansible/pull/69742

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -129,6 +129,20 @@ def __init__(self):
         self.resolved = False
+        self.plugin_type = None
+
+    @property
```
Vestigial?

s-hertel replied:
Thanks. Too many iterations of this thing :-)

nitzmahone commented on lib/ansible/plugins/loader.py:
```diff
@@ -593,6 +608,10 @@ def _resolve_plugin_step(self, name, mod_type='', ignore_deprecated=False,
                         plugin_load_context = self._find_fq_plugin(candidate_name, suffix, plugin_load_context=plugin_load_context)
+
+                        if candidate_name != plugin_load_context.original_name and candidate_name not in plugin_load_context.redirect_list:
+                            plugin_load_context.redirect_list.append(candidate_name)
```
This might be a redirection bug, lemme think on if I should be handling this elsewhere

nitzmahone commented on test/integration/targets/collections_action_groups/test_collections/ansible_collections/collection_namespace/collection_name/plugins/modules/example_module.py:
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/python
```
probably better to use an echo module that returns everything it was given- also need an echo action

s-hertel replied:
Fixed. I added two action plugins, one with the same name as the module and one with a different name.

nitzmahone commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -301,6 +301,11 @@ def parse(self, skip_action_validation=False):
                 action = item
                 thing = value
+
+                context = module_loader.find_plugin_with_context(item, check_aliases=True, collection_list=self._collection_list)
```
This doesn't cover actions and breaks `skip_action_validation=True` in some cases. It's doing the right thing, but you can deconstruct the `has_plugin` calls from conditional above into `find_plugin_with_context` calls instead, then hang onto the result and use it here.

nitzmahone reviewed (changes requested):
A few tweaks and a couple bugs (as we discussed), but looking good.

nitzmahone commented on lib/ansible/parsing/mod_args.py:
```diff
@@ -301,6 +301,11 @@ def parse(self, skip_action_validation=False):
+
+                context = module_loader.find_plugin_with_context(item, check_aliases=True, collection_list=self._collection_list)
+                if context.resolved and context.redirect_list:
+                    self._task_ds['ansible_internal_redirect_list'] = context.redirect_list
```
check w/ @jimi-c on this one- it definitely works, but dunno if there's a better way or if there are risks to this approach I'm unaware of...

s-hertel replied:
Improved this in 5e53832b683e20b202dededaccebffd37a45fb25

Thanks @bcoca and @jimi-c for the help!

s-hertel committed:
ci_complete

Fix a bug with action plugins and add tests

Remove vestigial code

s-hertel committed:
Remove task ds hack

jillr commented:
Latest patchset tests successfully locally against community.vmware, community.aws, and amazon.aws. That aspect of this LGTM.

s-hertel committed:
Always iterate over module defaults and evaluate any group since they may be redirected

Add a warning if redirections can't be loaded, but don't fail since module_defaults weren't previously validated

Fix tests

s-hertel committed:
Add action groups to ansible.builtin for incidental_ integration tests because ANSIBLE_LIBRARY is set to test/support/integration/plugins/modules

ci_complete

felixfontein commented:
@s-hertel should collections continue to include `action_groups_redirection` in meta/runtime.yml for now, or should we remove this until the future of this feature is clear?

felixfontein commented:
Also affects #69742 - should validation scream if `action_groups_redirection` is used, simply ignore it (and don't check its contents to make it possible to change it), or accept it only if it is in its current form?

AlanCoding commented:
I tested this with the proposal for the `awx.awx` collection. See my notes in that PR https://github.com/ansible/awx/pull/7391

This is working as the documentation in it advertises 👍 

The functionality seems necessary for basic parity with Ansible 2.9. Collections have been emboldened to add content at a faster rate than pre-migration, so this will become even more important for users.

AlanCoding commented on docs/docsite/rst/dev_guide/developing_collections.rst:
```diff
@@ -195,6 +197,132 @@ command completion, or environment variables are presented throughout the
+
+       module_defaults:
+         group/original_namespace.original_collection.original_group_name:
+           option_name: option_value
```
Could you also document how to use this inside of a role? I would be highly interested in that.

s-hertel commented:
@AlanCoding Thanks for taking a look at this. The portion of this PR responsible for backwards compatibility with 2.9 was moved to https://github.com/ansible/ansible/pull/69919 and merged for 2.10. I need to rebase since that was merged and I think it needs some more discussion.

AlanCoding commented:
Yep, all on the same page. It looks like #70232 was following the pattern established in #69919. I don't want to take a position on that, because this is _old content_ but the _action group_ is new. `awx.awx` can wait in line for this fix to solve the problem generally.

It is a useful new feature to add, because every module in this collection needs authentication parameters. When running locally, they take env vars or a config file. Neither of these are particularly clean when running on a _remote_ host, which is common to do where the remote host is the actual server (so it makes API calls to localhost). Besides those options `block:` will work, but indents everything.

AlanCoding commented:
Just a reminder that this would be awesome - but with heavy conflicts you may want to re-evaluate if it's worth keeping around.

s-hertel commented:
@AlanCoding Thanks for the reminder. Still want to work on this, but I agree it will be quicker to start from scratch at this point. :-)

felixfontein commented:
It's definitely awesome to be able to define action groups of the form `namespace.collection.action_group`.

I'm a bit wary about the feature to define short names in every collection, especially ones that are already in use by other collections. This would allow to create a `foo.bar` collection which has a useful module `left_pad` which coincidently grabs all module defaults, including cloud credentials of all kind, for various other modules and posts them to a random location on the web. (Well, I guess this can already happen now with all kind of plugins, but this would allow the same behavior for modules as well.)
